src/compiler/checker.ts

```diff
@@ -16771,37 +16780,69 @@ namespace ts {
             let match: Type | undefined;
-            propLoop: for (const prop of node.properties) {
+            for (const prop of node.properties) {
                 if (!prop.symbol) continue;
                 if (prop.kind !== SyntaxKind.PropertyAssignment) continue;
                 if (isPossiblyDiscriminantValue(prop.initializer) && isDiscriminantProperty(contextualType, prop.symbol.escapedName)) {
                     const discriminatingType = checkExpression(prop.initializer);
-                    for (const type of (contextualType as UnionType).types) {
-                        const targetType = getTypeOfPropertyOfType(type, prop.symbol.escapedName);
-                        if (targetType && isTypeAssignableTo(discriminatingType, targetType)) {
-                            if (match) {
-                                if (type === match) continue; // Finding multiple fields which discriminate to the same type is fine
-                                match = undefined;
-                                break propLoop;
-                            }
-                            match = type;
-                        }
+                    const result = getDiscriminationResultForProperty(contextualType, discriminatingType, prop.symbol.escapedName, match, isTypeAssignableTo);
+                    if (result === "mismatch") {
+                        return contextualType;
+                    }
+                    match = result;
+                }
+            }
+            return match || contextualType;
+        }
+
+        function discriminateContextualTypeByJSXAttributes(node: JsxAttributes, contextualType: UnionType) {
+            let match: Type | undefined;
+            for (const prop of node.properties) {
+                if (!prop.symbol) continue;
+                if (prop.kind !== SyntaxKind.JsxAttribute) continue;
+                let discriminatingType: Type | undefined;
+                let possiblyDiscriminant: boolean;
+                if (!prop.initializer) {
+                    discriminatingType = trueType;
+                    possiblyDiscriminant = !!isDiscriminantProperty(contextualType, prop.symbol.escapedName);
+                }
+                else {
+                    const simpleInitializer = isStringLiteral(prop.initializer) ? prop.initializer : prop.initializer.expression;
+                    possiblyDiscriminant = !!(simpleInitializer && isPossiblyDiscriminantValue(simpleInitializer) && isDiscriminantProperty(contextualType, prop.symbol.escapedName));
+                    if (possiblyDiscriminant) {
+                        discriminatingType = checkExpression(simpleInitializer!);
+                    }
+                }
+                if (possiblyDiscriminant) {
+                    const result = getDiscriminationResultForProperty(contextualType, discriminatingType!, prop.symbol.escapedName, match, isTypeAssignableTo);

```

Comment:
You could get rid of the nullable assertion if you dropped `possiblyDiscriminant` and checked `discriminatingType` above. The no-initialiser case would need to change to `discriminatingType = !!isDiscr... && trueType`.

Not sure if it's worth the change. Probably?